File: quickstart/javascript/deals-dashboard.mjs
// Tutorial 03 — A cheap-flights board: cheapest deal from one origin to many destinations.
// Run: RAPIDAPI_KEY=... RAPIDAPI_HOST=... node deals-dashboard.mjs [ORIGIN] [DEST,DEST,...]
import { TurismocityClient } from "./client.mjs";

const api = new TurismocityClient();
const [, , ORIGIN = "EZE", DESTS = "MAD,MIA,GRU,SCL,PUJ,MEX"] = process.argv;
const destinations = DESTS.split(",");

// Fetch all destinations in parallel, then sort by price.
const rows = await Promise.all(
  destinations.map(async (to) => {
    try {
      const d = await api.deals({ from: ORIGIN, to, currency: "USD" });
      const top = d.deals?.[0];
      return {
        to,
        cheapest: d.cheapest,
        roundtrip: d.byType.roundtrip,
        when: top ? `${top.outbound} → ${top.inbound ?? "one-way"}` : "",
        airline: top?.airlineName ?? "",
      };
    } catch {
      return { to, cheapest: null };
    }
  }),
);

rows.sort((a, b) => (a.cheapest ?? 1e9) - (b.cheapest ?? 1e9));

console.log(`\n✈  Cheapest flights from ${ORIGIN} (USD)\n`);
console.log("DEST   CHEAPEST   ROUND-TRIP   DATES                    AIRLINE");
console.log("-".repeat(72));
for (const r of rows) {
  if (r.cheapest == null) {
    console.log(`${r.to.padEnd(6)} (no data)`);
    continue;
  }
  console.log(
    `${r.to.padEnd(6)} $${String(r.cheapest).padEnd(8)} $${String(r.roundtrip).padEnd(10)} ` +
      `${(r.when || "").padEnd(24)} ${r.airline}`,
  );
}

File: quickstart/python/deals_dashboard.py
"""Tutorial 03 — A cheap-flights board: cheapest deal from one origin to many destinations.

Run: RAPIDAPI_KEY=... RAPIDAPI_HOST=... python deals_dashboard.py [ORIGIN] [DEST,DEST,...]
"""
import sys
from concurrent.futures import ThreadPoolExecutor
from client import TurismocityClient

# Print UTF-8 (airline names) correctly on Windows consoles too.
try:
    sys.stdout.reconfigure(encoding="utf-8")
except Exception:
    pass

api = TurismocityClient()
args = sys.argv[1:]
ORIGIN = args[0] if len(args) > 0 else "EZE"
DESTS = (args[1] if len(args) > 1 else "MAD,MIA,GRU,SCL,PUJ,MEX").split(",")


def fetch(to: str) -> dict:
    try:
        d = api.deals(to=to, currency="USD", **{"from": ORIGIN})
        top = (d.get("deals") or [None])[0]
        return {
            "to": to,
            "cheapest": d["cheapest"],
            "roundtrip": d["byType"]["roundtrip"],
            "when": f'{top["outbound"]} -> {top.get("inbound") or "one-way"}' if top else "",
            "airline": (top or {}).get("airlineName", ""),
        }
    except Exception:
        return {"to": to, "cheapest": None}


with ThreadPoolExecutor(max_workers=6) as pool:
    rows = list(pool.map(fetch, DESTS))

rows.sort(key=lambda r: r["cheapest"] if r["cheapest"] is not None else 1e9)

print(f"\nCheapest flights from {ORIGIN} (USD)\n")
print("DEST   CHEAPEST   ROUND-TRIP   DATES                    AIRLINE")
print("-" * 72)
for r in rows:
    if r["cheapest"] is None:
        print(f'{r["to"]:<6} (no data)')
        continue
    print(
        f'{r["to"]:<6} ${str(r["cheapest"]):<8} ${str(r["roundtrip"]):<10} '
        f'{r["when"]:<24} {r["airline"]}'
    )
